src/logger/src/logger.rs

```diff
@@ -1,776 +1,343 @@
-//! 2018-11-07T05:34:25.180751152 [anonymous-instance:ERROR:vmm/src/lib.rs:1173] Failed to write
-//! metrics: Failed to write logs. Error: operation would block
-//! ```
-//! # Limitations
-//! Logs can be flushed either to stdout/stderr or to a byte-oriented sink (File, FIFO, Ring Buffer
-//! etc).
-
-use std::fmt::{self, Debug};
-use std::io::{sink, stdout, Write};
-use std::sync::atomic::{AtomicBool, Ordering};
-use std::sync::{Mutex, RwLock};
-use std::{result, thread};
-
-use lazy_static::lazy_static;
-use log::{max_level, set_logger, set_max_level, LevelFilter, Log, Metadata, Record};
+use std::fmt::Debug;
+use std::io::Write;
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::{Mutex, OnceLock};
+use std::thread;
+
+use log::{Log, Metadata, Record};
+use serde::{Deserialize, Serialize};
 use utils::time::LocalTime;
 
-use super::extract_guard;
-use crate::init;
-use crate::init::Init;
 use crate::metrics::{IncMetric, METRICS};
 
-/// Type for returning functions outcome.
-pub type Result<T> = result::Result<T, LoggerError>;
-
-// Values used by the Logger.
-const DEFAULT_MAX_LEVEL: LevelFilter = LevelFilter::Warn;
-
-lazy_static! {
-    static ref _LOGGER_INNER: Logger = Logger::new();
-
-    /// Static instance used for handling human-readable logs.
-    pub static ref LOGGER: &'static Logger = {
-        set_logger(_LOGGER_INNER.deref()).expect("Failed to set logger");
-        _LOGGER_INNER.deref()
-    };
-}
-
-/// Logger representing the logging subsystem.
-// All member fields have types which are Sync, and exhibit interior mutability, so
-// we can call logging operations using a non-mut static global variable.
-pub struct Logger {
-    init: Init,
-    // Human readable logs will be outputted here.
-    log_buf: Mutex<Box<dyn Write + Send>>,
-    show_level: AtomicBool,
-    show_file_path: AtomicBool,
-    show_line_numbers: AtomicBool,
-    instance_id: RwLock<String>,
-}
-
-impl fmt::Debug for Logger {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Logger")
-            .field("init", &self.init)
-            .field("log_buf", &"?")
-            .field("show_level", &self.show_level)
-            .field("show_file_path", &self.show_file_path)
-            .field("show_line_numbers", &self.show_line_numbers)
-            .field("instance_id", &self.instance_id)
-            .finish()
-    }
-}
+/// Default instance id.
+pub const DEFAULT_INSTANCE_ID: &str = "anonymous-instance";
+/// Instance id.
+pub static INSTANCE_ID: OnceLock<String> = OnceLock::new();
+
+/// The logger.
+///
+/// Default values matching the swagger specification (`src/api_server/swagger/firecracker.yaml`).
+pub static LOGGER: Logger = Logger(Mutex::new(LoggerConfiguration {
+    target: None,
+    filter: LogFiler {
+        level: log::LevelFilter::Warn,
+    },
+    format: LogFormat {
+        show_level: false,
+        show_log_origin: false,
+    },
+}));
```

Comment:
Mhh, this adds more locking that I don't think its necessary (but concurrency in Rust is also not my strongest suit). But currently we need to lock the entire logger if we check the level (in enabled below). This means that even if a log message does not get printed, we will do a `Mutex::lock`. Cant we instead only wrap the `target` into a `Mutex`, and have the log level be a `AtomicU8` and `show_level` and `show_log_origin` be `AtomicBool`? They allow interior mutability through  `store`, and atomics should be more performant than mutexes (at the very least they should match whatever performance we have currently, as the old logger uses atomics for these)

Reply:
There will still need to be a `Mutex` wrapping `target` by having a `Mutex` on `show_level` and `show_log_origin` it prevents these being updated mid logging some data (and is a bit simpler).

Moving this `Mutex` inner would only give a speedup if someone is very frequently updating the logger format. I don't think this case is common.

Reply:
You are right, I wrote that comment under the assumption that the mutex would also need to be held to determine whether a message should be printed in the first place. If that is not the case (e.g. mutex only needs to be held when a message is actually logged), then there's no big difference, and this does indeed make it slightly simpler